core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -764,7 +762,7 @@ public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Expression
             PlanNode output = node;
             if (leftSource != node.getLeft() ||
                     rightSource != node.getRight() ||
-                    !areExpressionsEquivalent(newJoinPredicate, joinPredicate)) {
+                    !newJoinPredicate.equals(joinPredicate)) {
```

Comment:
so the assumption is that newJoinPredicate cannot be similar to (equivalent), but different from, joinPredicate?

what does guarantee this?

Reply:
This is just a stronger requirement to detect whether the predicate was rewritten by the optimization. Before, it just checked whether they were equivalent based on some semantic-preserving transformations. Now it's based on whether they are exactly the same.

Reply:
The question is, was `ExpressionEquivalence` more forgiving than `equals`? Maybe we keep `ExpressionEquivalence` abstraction but with `equals` inside?

Reply:
Yes, it was more lenient than equality, however, I don't think that matters in practice.

The only canonicalizations this class supports are:

* `AND(AND(x, y), z)` => `AND(x, y, z)`. Same for `OR`.
* `AND(x, x, x, a)` => `AND(x, a)`. Same for `OR`. I believe there's a bug here, though, as it ignores non-deterministic operations that should be evaluated multiple times.
* `a = b` => `b = a`
* `a is distinct from b` => `b is distinct from a`

All of these are already handled by the expression optimizer and canonicalizer.

Reply:
> I believe there's a bug here, though, as it ignores non-deterministic operations that should be evaluated multiple times.

This demonstrates the issue:

```java
ResolvedFunction random = new TestingFunctionResolution().resolveFunction("random", fromTypes());
Expression expression = new IsNull(new Call(random, List.of()));
assertNotEquivalent(
        new Logical(AND, List.of(expression, expression)),
        expression);
```

Reply:
> All of these are already handled by the expression optimizer and canonicalizer.

the `newJoinPredicate` comes thru the optimizer (`PredicatePushDown.Rewriter#simplifyExpression`), does it also do canonicalization?